### PR TITLE
Implements Backup Improvements for 3.4

### LINKED ISSUE #756: Backup improvements (for 3.4)
- [ ] Use new backup system for android versions
- [ ] Use new backup system for backuping documents too
- [ ] When saving backup, location is not useful. (due to #748)

Screenshot_1593688375

Screenshot_1593688434

### PR DESCRIPTION
**Pull request by RuggMatt**
- [x] Users on all Android versions are now asked where they want to save the file when picking the phone storage option from the backup system
- [x] When backing up modules, users now have an option to choose phone storage
- [x] When a backup is completed, the location no longer shows
- [x] Removed unnecessary WRITE_EXTERNAL_STORAGE permissions since it is no longer being used to directly write to external storage.

Closes #756 

**Describe the pull request content**
A clear and concise description of what the pull request is about.
What are the benefits of this new code addition? 
Possible negative side effects? 

**Screenshots**
If applicable, add screenshots to help explain your pull request.

### CODE REVIEW

**Comment by tuomas2** [comment]
After merging other PRs, there are now little merge conflicts

**Comment by RuggMatt** [comment]
Should be good to go now. Also added a commit to remove some of the now unused code...

File: app/src/main/java/net/bible/android/control/backup/BackupControl.kt
Comment (JJK96) [inline app/src/main/java/net/bible/android/control/backup/BackupControl.kt:116]: Why is this removed? I would think it is useful to see where the file was saved.

File: app/src/main/java/net/bible/android/view/activity/page/MenuCommandHandler.kt
Comment (JJK96) [inline app/src/main/java/net/bible/android/view/activity/page/MenuCommandHandler.kt:223]: Since this is removed, I think this needs testing on those android versions.

File: app/src/main/java/net/bible/android/view/activity/page/MenuCommandHandler.kt
Comment (JJK96) [inline app/src/main/java/net/bible/android/view/activity/page/MenuCommandHandler.kt:262]: Since this is removed, I think this needs testing on those android versions.

**Review by JJK96 — CHANGES_REQUESTED** [review]
I've reviewed the PR. 

It no longer gives an error when trying to save on Android 10, so that's a big improvement already!
I tested saving and restoring the application database and modules to/from phone storage which all worked fine on Android 10.

I ran into one problem, backup to phone storage does not overwrite an existing file. 
So I end up with `andBibleDatabase.db`, `andBibleDatabase.db (1)` and `andBibleDatabase.db (2)`.
That issue has to be resolved before merging, IMO.

File: app/src/main/java/net/bible/android/control/backup/BackupControl.kt
Comment (tuomas2) [reply app/src/main/java/net/bible/android/control/backup/BackupControl.kt:116]: Check out screenshots on #756.

Comment (JJK96) [reply app/src/main/java/net/bible/android/control/backup/BackupControl.kt:116]: I missed that issue, good change then

File: app/src/main/java/net/bible/android/view/activity/page/MenuCommandHandler.kt
Comment (tuomas2) [reply app/src/main/java/net/bible/android/view/activity/page/MenuCommandHandler.kt:223]: It would indeed be good to roughly go through different android versions with emulator (from lollipop (5) to the latest models).

**Comment by RuggMatt** [comment]
> 
> 
> I've reviewed the PR.
> 
> It no longer gives an error when trying to save on Android 10, so that's a big improvement already!
> I tested saving and restoring the application database and modules to/from phone storage which all worked fine on Android 10.
> 
> I ran into one problem, backup to phone storage does not overwrite an existing file.
> So I end up with `andBibleDatabase.db`, `andBibleDatabase.db (1)` and `andBibleDatabase.db (2)`.
> That issue has to be resolved before merging, IMO.

Regarding the multiple file names, this is an android file level problem as the intent automatically creates a new file if one with the same name already exists. 

Possible options are to show another screen when choosing backup when user clicks phone storage as described below

 #### Current: 
When a user selects the **phone storage** option:
  - show file picker and write to file
 #### Possible Change:
When a user selects the **phone storage** option
- if it is the user's first time backing up to phone storage
   - ask them where to create a new file
   - write backup to file
   - call [`ContentResolver#takePersistableUriPermission(Uri, int)`](https://developer.android.com/reference/android/content/ContentResolver#takePersistableUriPermission\(android.net.Uri,%20int\)) to keep write permissions to file across phone reboots
   - possibly need to save URI somewhere or figure out how to use [`ContentResolver#getPersistedUriPermissions()`](https://developer.android.com/reference/android/content/ContentResolver#getPersistedUriPermissions\(\))
- else the user has created a backup before
  - ask them if they want to overwrite the old backup file or create a new one?
    - overwrite?
      - retrieve old backup Uri using `getPersistedUriPermissions()` and figure out which one is the correct type of backup?
      - write backup to Uri
      - maybe a menu so a user can select which backup to overwrite if user has created multiple different backups before
    - or create a new file
      - user is prompted to create a new file through the android UI file picker
      - write backup to file
      - `takePersistableUriPermission...`

**Comment by JJK96** [comment]
I like the simplicity of your current solution. I think this is a problem that should be solved by Android indeed.

So, considering that the issue is not easily resolved at the moment, and the current status is an improvement with respect to the previous version, this PR can be merged IMO (After some testing on older android versions).

**Comment by RuggMatt** [comment]
was doing some testing on older versions and i discovered this error in general.. 
When a user backs up the database after opening the app, no "success" message is shown. 
If a user immediately backs up again, the "success" message works as normal. 
```
E/Dialogs: Error showing error message.  Original error msg:My Notes, Bookmarks, Reading Plans and Workspaces backed up successfully.
    java.lang.RuntimeException: Can't create handler inside thread that has not called Looper.prepare()
        at android.os.Handler.<init>(Handler.java:200)
        at android.os.Handler.<init>(Handler.java:114)
        at android.widget.Toast$TN.<init>(Toast.java:344)
        at android.widget.Toast.<init>(Toast.java:100)
        at android.widget.Toast.makeText(Toast.java:258)
        at net.bible.android.view.activity.base.Dialogs.showMsg(Dialogs.kt:119)
        at net.bible.android.view.activity.base.Dialogs.showErrorMsg(Dialogs.kt:87)
        at net.bible.android.view.activity.base.Dialogs.showErrorMsg$default(Dialogs.kt:86)
        at net.bible.android.view.activity.base.Dialogs.showMsg(Dialogs.kt:51)
        at net.bible.android.control.backup.BackupControl.backupDatabaseToUri(BackupControl.kt:96)
        at net.bible.android.view.activity.page.MainBibleActivity$onActivityResult$2.invokeSuspend(MainBibleActivity.kt:1158)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```
Does anyone know how to fix?

**Comment by timbze** [comment]
> Does anyone know how to fix?

I great to see this stuff being tested thoroughly. :) This issue is because the dialog is trying to start on a background thread, but it must be started on the main thread.

In `MainBibleActivity.kt`, L1133+ (master branch)
```kotlin
GlobalScope.launch(Dispatchers.IO) {
  val inputStream = contentResolver.openInputStream(data!!.data!!)
  if (backupControl.restoreDatabaseViaIntent(inputStream!!)) {
    windowControl.windowSync.setResyncRequired()
    withContext(Dispatchers.Main) {
      documentViewManager.clearBibleViewFactory()
      currentWorkspaceId = 0
    }
  }
}
```

`restoreDatabaseViaIntent` is being started under Dispatchers.IO, and that is where the error occurs. It either needs to start in Dispatchers.Main, or the Dialogs.instance.showMsg within `restoreDatabaseViaIntent` needs to be within another `GlobalScope.launch(Dispatchers.Main)`

**Comment by timbze** [comment]
@mattschlosser can you either merge or rebase develop into your PR so that CI can pass? There was a CI issue that has been fixed.

**Comment by JJK96** [comment]
I just tested the new version, it works fine for me. 
In the previous version I could no longer choose a filename for the backup to phone storage, but that is fixed in this version.

File: app/src/main/res/values/strings.xml
Comment (tuomas2) [inline app/src/main/res/values/strings.xml:484]: Because of translations, we should not make such changes to strings that change parameters. Otherwise users with translations that have not been updated will have crashes.

When parameters need to be changed, like here, we need to introduce a new string (with new key, for example backup_success2), use it, and remove old string.

**Comment by tuomas2** [comment]
Looks okay. I will do the fix related to my own comment in develop branch. Let's continue testing there.

**Pull request merged**
